proposals/testing/NNNN-attachments.md

```diff
@@ -0,0 +1,436 @@
+/// written to disk when a test is run.
+///
+/// To attach an attachable value to a test report or test run output, use it to
+/// initialize a new instance of ``Attachment``, then call
+/// ``Attachment/attach(sourceLocation:)``. An attachment can only be attached
+/// once.
+///
+/// The testing library provides default conformances to this protocol for a
+/// variety of standard library types. Most user-defined types do not need to
+/// conform to this protocol.
+///
+/// A type should conform to this protocol if it can be represented as a
+/// sequence of bytes that would be diagnostically useful if a test fails. If a
+/// type cannot conform directly to this protocol (such as a non-final class or
+/// a type declared in a third-party module), you can create a container type
+/// that conforms to ``AttachableContainer`` to act as a proxy.
+public protocol Attachable: ~Copyable {
+  /// An estimate of the number of bytes of memory needed to store this value as
+  /// an attachment.
+  ///
+  /// The testing library uses this property to determine if an attachment
+  /// should be held in memory or should be immediately persisted to storage.
+  /// Larger attachments are more likely to be persisted, but the algorithm the
+  /// testing library uses is an implementation detail and is subject to change.
+  ///
+  /// The value of this property is approximately equal to the number of bytes
+  /// that will actually be needed, or `nil` if the value cannot be computed
+  /// efficiently. The default implementation of this property returns `nil`.
+  ///
+  /// - Complexity: O(1) unless `Self` conforms to `Collection`, in which case
+  ///   up to O(_n_) where _n_ is the length of the collection.
+  var estimatedAttachmentByteCount: Int? { get }
+
+  /// Call a function and pass a buffer representing this instance to it.
+  ///
+  /// - Parameters:
+  ///   - attachment: The attachment that is requesting a buffer (that is, the
+  ///     attachment containing this instance.)
+  ///   - body: A function to call. A temporary buffer containing a data
+  ///     representation of this instance is passed to it.
+  ///
+  /// - Returns: Whatever is returned by `body`.
+  ///
+  /// - Throws: Whatever is thrown by `body`, or any error that prevented the
+  ///   creation of the buffer.
+  ///
+  /// The testing library uses this function when writing an attachment to a
+  /// test report or to a file on disk. The format of the buffer is
+  /// implementation-defined, but should be "idiomatic" for this type: for
+  /// example, if this type represents an image, it would be appropriate for
+  /// the buffer to contain an image in PNG format, JPEG format, etc., but it
+  /// would not be idiomatic for the buffer to contain a textual description of
+  /// the image.
+  borrowing func withUnsafeBytes<R>(for attachment: borrowing Attachment<Self>, _ body: (UnsafeRawBufferPointer) throws -> R) throws -> R
```

Comment:
Is there a reason for the closure parameter being an unsafe type as opposed to, say, a `RawSpan`?

Reply:
Yes, as discussed in "Future Directions", Swift Testing needs to back-deploy earlier than `RawSpan` is available, and various types we want to make attachable (e.g. `Data`) don't have the capacity to produce a `RawSpan` yet.